app/services/puppet_fact_parser.rb

```diff
@@ -190,48 +197,127 @@ def use_legacy_facts?
     os_name = facts.dig(:os, :name).presence || facts[:operatingsystem].presence || raise(::Foreman::Exception.new("invalid facts, missing operating system value"))
 
-    if os_name == 'RedHat' && facts[:lsbdistid] == 'RedHatEnterpriseWorkstation'
+    if os_name == 'RedHat' && distro_id == 'RedHatEnterpriseWorkstation'
       os_name += '_Workstation'
     end
 
     os_name
+  rescue ::Foreman::Exception
+    raise
+  rescue StandardError => e
+    logger.error { "Failed to read the OS name: #{e}" }
+    raise(::Foreman::Exception.new("invalid facts, missing operating system value"))
   end
 
   def os_release
     case os_name
-    when /(suse|sles|gentoo)/i
-      facts[:operatingsystemrelease]
     when /(windows)/i
       facts[:kernelrelease]
     when /AIX/i
-      majoraix, tlaix, spaix, _yearaix = facts[:operatingsystemrelease].split("-")
+      majoraix, tlaix, spaix, _yearaix = os_release_full.split("-")
       majoraix + "." + tlaix + spaix
     when /JUNOS/i
-      majorjunos, minorjunos = facts[:operatingsystemrelease].split("R")
+      majorjunos, minorjunos = os_release_full.split("R")
       majorjunos + "." + minorjunos
     when /FreeBSD/i
-      facts[:operatingsystemrelease].gsub(/\-RELEASE\-p[0-9]+/, '')
+      os_release_full.gsub(/\-RELEASE\-p[0-9]+/, '')
     when /Solaris/i
-      facts[:operatingsystemrelease].gsub(/_u/, '.')
+      os_release_full.gsub(/_u/, '.')
     when /PSBM/i
-      majorpsbm, minorpsbm = facts[:operatingsystemrelease].split(".")
+      majorpsbm, minorpsbm = os_release_full.split(".")
       majorpsbm + "." + minorpsbm
     when /Archlinux/i
       # Archlinux is a rolling release, so it has no releases. 1.0 is always used
       '1.0'
     when /Debian/i
-      return "99" if facts[:lsbdistcodename] =~ /sid/
-      release = facts.dig(:os, :release, :full) || facts[:lsbdistrelease] || facts[:operatingsystemrelease]
+      return "99" if distro_codename =~ /sid/
+      release = os_release_full
       case release
       when 'bullseye/sid' # Debian Bullseye testing will be 11
         '11'
       else
         release
       end
     else
-      facts.dig(:os, :release, :full) || facts[:lsbdistrelease] || facts[:operatingsystemrelease]
+      os_release_full
     end
   end
+
+  # The full OS release (7 / 7.9 / 7.6.1810 / 2012 R2 / 20.04)
+  def os_release_full
+    # Facter 2.2 introduced the os.release fact
+    facts.dig(:os, :release, :full) || facts[:operatingsystemrelease]
+  rescue StandardError => e
+    logger.error { "Failed to read the full OS release: #{e}" }
+    nil
+  end
+
+  # This fact returns the distribution's id, which typically relies on
+  # lsb-release to be installed. As such, it's an optional fact
+  def distro_id
+    # Facter 3.0 introduced the os.distro fact
+    facts.dig(:os, :distro, :id).presence || facts[:lsbdistid].presence
+  rescue StandardError => e
+    logger.warning { "Failed to the read distribution id: #{e}" }
+    nil
+  end
```

Comment:
All those rescue blocks are useless, dig or presence will either return something or nil.

Reply:
This is false. It was the entire conclusion of https://community.theforeman.org/t/puppet-fact-parser-is-failing-with-string-does-not-have-dig-method/23138

The tl;dr of that was that for compatibility with older code Puppet can stringify fact values. That means the `os` structure is stringified and you can't dig into that.

The minimal reproducer:
```
[1] pry(main)> {'os' => 'some string'}.dig('os', 'name')
TypeError: String does not have #dig method
from (pry):1:in `dig'
```